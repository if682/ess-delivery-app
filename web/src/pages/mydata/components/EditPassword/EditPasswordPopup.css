.input-texts{
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
}

.edit-password-text{
    padding-left: 3rem;
    padding-bottom: 0.5rem;
    font-style: normal;
    font-weight: 400;
    font-size: 16px;
    line-height: 125%;

    display: flex;
    align-items: center;
    text-align: center;

    color: #424141;
}