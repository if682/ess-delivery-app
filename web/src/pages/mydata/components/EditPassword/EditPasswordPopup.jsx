import Modal from 'react-bootstrap/Modal';
import Button from 'react-bootstrap/Button';
import { useEffect, useState } from 'react';
import { isInputNull } from '../../../../shared/functions/isInputNull';
import { WarningText } from '../../../../components/atoms/warning-text/WarningText';
import { InputText } from '../../../../components/atoms/input-text/InputText';
import './EditPasswordPopup.css'

export const EditPasswordPopup = (props) => {
    const [password, setPassword] = useState("");
    const [oldPassword, setOldPassword] = useState("")
    const [warningMessage, setWarningMessage] = useState(null);

    //warningMessage recebe valor nulo quando os valores da senha antiga e da nova são atualizados
    useEffect(() => {
        setWarningMessage(null)
    }, [password, oldPassword])


    async function EditPassword () {
        //verifica se há algum input com valor nulo
        if(isInputNull(password) || isInputNull(oldPassword)){
            setWarningMessage('Preencha todos os campos!')
        }
        //checa se o valor fornecido para oldPassword corresponde a senha atual do cliente (validação)
        else if(oldPassword != props.currentPassword){
            setWarningMessage('O valor fornecido para a senha atual está incorreto')
        }
        else {
            let item = {name: props.name, email: props.email, password: password}

            //Atualiza o valor da senha do cliente
            await fetch(`http://localhost:3001/clients/${props.client_id}`, {
                method: 'PUT',
                headers: {
                'Content-Type': 'application/json'
                },
                body: JSON.stringify(item)
            });
        }
    }

    return (
        <Modal
        {...props}
        size="md"
        aria-labelledby="contained-modal-title-vcenter"
        centered
        >
        <Modal.Header closeButton {...props}>
            <Modal.Title id="contained-modal-title-vcenter">
            Alterar senha
            </Modal.Title>
        </Modal.Header>
        <Modal.Body>
            <text className='edit-password-text'>
            Digite sua senha atual e logo em seguida a nova senha
            </text>
            <div className='input-texts'>
                <InputText setInput={setOldPassword} type="password" placeholder="Senha atual"/>
                <InputText setInput={setPassword} type="password" placeholder="Nova senha"/>
                <WarningText warningMessage={warningMessage}/>
            </div>
        </Modal.Body>
        <Modal.Footer>
            <Button className='btn-secondary' onClick={props.onHide}>Cancelar</Button>
            <Button className='btn-primary' onClick={() => EditPassword()}>Salvar</Button>
        </Modal.Footer>
        </Modal>
    );
}
