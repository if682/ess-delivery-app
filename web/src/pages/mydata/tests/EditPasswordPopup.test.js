import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { EditPasswordPopup } from "../components/EditPassword/EditPasswordPopup";

describe("EditPasswordPopup", () => {
    let onHideMock, props;
    
    //mocka a função onHide e o props recebidos pelo popup
    beforeEach(() => {
        onHideMock = jest.fn();
        props = {
            name: 'teste',
            email: 'teste@hotmail.com',
            client_id: 123,
            currentPassword: 'oldpassword',
            show: true,
            onHide: onHideMock
        };
    });
    
    //Verifica se o popup é renderizado corretamente 
    it('renders correctly', () => {
        const { container } = render(<EditPasswordPopup {...props} />);

       //checa se o que foi renderizado corresponde ao snapshot do componente
       expect(container.firstChild).toMatchSnapshot();
    });

//Verifica se a função mockada onHide é chamada quando se clica no botão de cancelar
    it('should call onHide when cancel button is clicked', () => {
        const { getByText } = render(<EditPasswordPopup {...props} />);
        fireEvent.click(getByText('Cancelar'));
        expect(onHideMock).toHaveBeenCalled();
      });

//Verifica se o popup expõe a mensagem de erro "Preencha todos os campos!" se houver alguma caixa de texto vazia
    it('should display warning message "Preencha todos os campos!" if either password or oldPassword is null', async () => {
        const {getByPlaceholderText, getByText} = render(<EditPasswordPopup {...props} />);
        const oldPasswordField = getByPlaceholderText("Senha atual");
        const passwordField = getByPlaceholderText("Nova senha");
        const saveButton = getByText("Salvar");

        fireEvent.change(oldPasswordField, { target: { value: "a" } });
        fireEvent.change(passwordField, { target: { value: "" } });
        fireEvent.click(saveButton);

        const warningMessage = await screen.findByText("Preencha todos os campos!");
        expect(warningMessage).toBeInTheDocument();
    });

//Verifica se o popup expõe a mensagem de erro "O valor fornecido para a senha atual está incorreto" quando o valor recebido como senha não coincide com o valor da senha atual
    it('should display the warning message "O valor fornecido para a senha atual está incorreto" if oldPassword is incorrect', async () => {
        const {getByPlaceholderText, getByText} = render(<EditPasswordPopup {...props} />);
        const oldPasswordField = getByPlaceholderText("Senha atual");
        const passwordField = getByPlaceholderText("Nova senha");
        const saveButton = getByText("Salvar");

        fireEvent.change(oldPasswordField, { target: { value: "wrongPassword" } });
        fireEvent.change(passwordField, { target: { value: "newPassword" } });
        fireEvent.click(saveButton);

        const warningMessage = await screen.findByText(
        "O valor fornecido para a senha atual está incorreto"
        );
        expect(warningMessage).toBeInTheDocument();
    });

    //Verifica se a função fetch é chamada para atualizar a senha do usuário quando todas as informações recebidas no input são válidas
    it("should call the fetch function to update the password when save button is clicked with correct inputs", async () => {
        global.fetch = jest.fn(() =>
        Promise.resolve({
            json: () => Promise.resolve({ id: 123, name: 'test', email: 'teste@hotmail.com', password: 'newPassword' })
        })
        );
        const { getByText, getByPlaceholderText} = render(<EditPasswordPopup {...props} />);
        const oldPasswordField = getByPlaceholderText("Senha atual");
        const passwordField = getByPlaceholderText("Nova senha");

        fireEvent.change(oldPasswordField, { target: { value: "oldpassword" } });
        fireEvent.change(passwordField, { target: { value: "newPassword" } });
        fireEvent.click(getByText("Salvar"));
        await waitFor(() => expect(fetch).toHaveBeenCalledWith('http://localhost:3001/clients/123', {
        method: 'PUT',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify({ name: 'teste', email: 'teste@hotmail.com', password: 'newPassword'})
        }));

    });
});
