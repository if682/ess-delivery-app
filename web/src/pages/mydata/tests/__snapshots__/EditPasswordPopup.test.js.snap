// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditPasswordPopup renders correctly 1`] = `null`;
